cardano-ledger-test/src/Test/Cardano/Ledger/ModelChain.hs

```diff
@@ -0,0 +1,698 @@
+
+class HasEraElaboratorState s era | s -> era where
+  eraElaboratorState
+    :: forall f. Functor f
+    => (EraElaboratorState era -> f (EraElaboratorState era)) -> s -> f s
+
+instance HasEraElaboratorState (EraElaboratorState era) era where
+  eraElaboratorState = id
+
+eesUnusedKeyPairs :: Functor f => (Word64 -> f Word64) -> EraElaboratorState era -> f (EraElaboratorState era)
+eesUnusedKeyPairs a2fb s = (\b -> s {_eesUnusedKeyPairs = b}) <$> a2fb (_eesUnusedKeyPairs s)
+
+eesKeys :: Functor f => (Map.Map ModelAddress (TestKeyPair (Crypto era)) -> f (Map.Map ModelAddress (TestKeyPair (Crypto era)))) -> EraElaboratorState era -> f (EraElaboratorState era)
+eesKeys a2fb s = (\b -> s {_eesKeys = b}) <$> a2fb (_eesKeys s)
+
+eesTxIds :: Functor f => (Map.Map ModelTxId (Shelley.TxId (Crypto era)) -> f (Map.Map ModelTxId (Shelley.TxId (Crypto era)))) -> EraElaboratorState era -> f (EraElaboratorState era)
+eesTxIds a2fb s = (\b -> s {_eesTxIds = b}) <$> a2fb (_eesTxIds s)
+
+eesCurrentEpoch :: Functor f => (EpochNo -> f EpochNo) -> EraElaboratorState era -> f (EraElaboratorState era)
+eesCurrentEpoch a2fb s = (\b -> s {_eesCurrentEpoch = b}) <$> a2fb (_eesCurrentEpoch s)
+
+eesPendingWitnessKeys :: Functor f => ([KeyPair 'Witness (Crypto era)] -> f [KeyPair 'Witness (Crypto era)]) -> EraElaboratorState era -> f (EraElaboratorState era)
+eesPendingWitnessKeys a2fb s = (\b -> s {_eesPendingWitnessKeys = b}) <$> a2fb (_eesPendingWitnessKeys s)
+
+eesUTxOs :: Lens' (EraElaboratorState era) (Map.Map ModelUTxOId ModelAddress)
+eesUTxOs a2fb s = (\b -> s {_eesUTxOs = b}) <$> a2fb (_eesUTxOs s)
+
+eesCurrentSlot :: Lens' (EraElaboratorState era) SlotNo
+eesCurrentSlot a2fb s = (\b -> s {_eesCurrentSlot = b}) <$> a2fb (_eesCurrentSlot s)
+
+data TestKeyPair crypto = TestKeyPair
+  { _tkpKeyPair :: KeyPair' crypto
+  , _tkpVRF :: (SignKeyVRF crypto , VerKeyVRF crypto )
+  , _tkpAddr :: Addr crypto
+  , _tkpVRFHash :: Hash.Hash (C.HASH crypto) (Cardano.Crypto.VRF.Class.VerKeyVRF (C.VRF crypto))
+  , _tkpStakePool :: KeyHash 'Staking crypto
+  , _tkpStakeCredential :: StakeCredential crypto
+  }
+
+deriving instance
+  ( C.Crypto crypto
+  )
+  => Show (TestKeyPair crypto)
+
+
+getKeyPairForImpl
+  :: forall m era st proxy.
+  ( MonadState st m
+  , HasEraElaboratorState st era
+  , C.Crypto (Crypto era)
+  )
+  => proxy era -> ModelAddress
+  -> m (TestKeyPair (Crypto era))
+getKeyPairForImpl _ mAddr = do
+  st <- use eraElaboratorState
+  case Map.lookup mAddr (_eesKeys st) of
+    Just k -> pure k
+    Nothing -> do
+      unusedKeyPairId <- eraElaboratorState . eesUnusedKeyPairs <<%= succ
+
+      let
+        keyPair@(_, poolKey) = mkKeyPairs @(Crypto era) unusedKeyPairId
+        vrf@(_, vrf') = mkVRFKeyPair @(C.VRF (Crypto era)) (1, 0, 0, 0, unusedKeyPairId)
+        k = TestKeyPair
+          { _tkpKeyPair = keyPair
+          , _tkpVRF = vrf
+          , _tkpAddr = toAddr Testnet keyPair -- TODO: network can be read from Globals
```

Comment:
it's fine by me to leave this hard coded as `Testnet`.

Reply:
My thinking was that the network is in Globals; and getting it from context is a way to avoid some latent bugs.

Reply:
yea, that's reasonable